core-bundle/src/Picker/UniversalPickerProvider.php

```diff
@@ -0,0 +1,267 @@
+
+        Controller::loadDataContainer($table);
+
+        $pid = null;
+        $ptable = $GLOBALS['TL_DCA'][$table]['config']['ptable'] ?? null;
+        $dynamicPtable = $GLOBALS['TL_DCA'][$table]['config']['dynamicPtable'] ?? false;
+
+        if ($ptable || $dynamicPtable) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->select('pid')->from($table)->where($qb->expr()->eq('id', $value));
+
+            if ($dynamicPtable) {
+                $qb->addSelect('ptable');
+            }
+
+            $data = $qb->execute()->fetch();
+
+            if (false === $data) {
+                return $this->getUrlForValue($config, array_keys($modules)[0]);
+            }
+
+            $pid = $data['pid'];
+
+            if ($dynamicPtable) {
+                $ptable = $data['ptable'] ?: $ptable;
+
+                // Backwards compatibility, assuming old data in tl_content
+                if (!$ptable) {
+                    $ptable = 'tl_article';
+                }
+            }
+        }
+
+        if (1 === \count($modules) || !$ptable) {
+            return $this->getUrlForValue($config, array_keys($modules)[0]);
+        }
+
+        foreach ($modules as $module => $tables) {
+            if (\in_array($ptable, $tables, true)) {
+                return $this->getUrlForValue($config, $module, $table, $pid);
+            }
+        }
+
+        return $this->getUrlForValue($config, array_keys($modules)[0], $table, $pid);
+    }
+
+    public function addMenuItems(ItemInterface $menu, PickerConfig $config): void
+    {
+        $modules = array_keys($this->getModulesForTable($config->getContext()));
+
+        foreach ($modules as $name) {
+            $params = array_merge(
+                [
+                    'do' => $name,
+                    'popup' => '1',
+                    'picker' => $config->cloneForCurrent('universal.'.$name)->urlEncode(),
+                ]
+            );
+
+            $menu->addChild($this->menuFactory->createItem(
+                $name,
+                [
+                    'label' => $this->translator->trans('MOD.'.$name.'.0', [], 'contao_default'),
+                    'linkAttributes' => ['class' => $name],
+                    'current' => $this->isCurrent($config) && substr($config->getCurrent(), 10) === $name,
+                    'uri' => $this->router->generate('contao_backend', $params),
+                ]
+            ));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createMenuItem(PickerConfig $config): ItemInterface
+    {
+        $menu = $this->menuFactory->createItem('picker');
+
+        $this->addMenuItems($menu, $config);
+
+        return $menu->getFirstChild();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsContext($context): bool
+    {
+        return 0 === strpos($context, 'universal.') && 0 !== \count($this->getModulesForTable($context));
```

Comment:
`universal.` should probably be a private class constant :)

Reply:
Refactored in 319ef87f32e7fc220b67778d566e6a6e8fd91b35